library/std/src/thread/local.rs

```diff
@@ -1028,10 +1025,16 @@ pub mod fast {
         // `Option<T>` to `None`, and `dtor_state` to `RunningOrHasRun`. This
         // causes future calls to `get` to run `try_initialize_drop` again,
         // which will now fail, and return `None`.
-        unsafe {
+        //
+        // Wrap the call in a catch to ensure unwinding is caught in the event
+        // a panic takes place in a destructor.
+        if let Err(_) = panic::catch_unwind(panic::AssertUnwindSafe(|| unsafe {
+            let ptr = ptr as *mut Key<T>;
```

Comment:
Is `panic::AssertUnwindSafe` correct here? Is it allowed to move `let ptr = ` down to limit its scope?

Reply:
You only have to put the `drop()` line in the catch_unwind.

Reply:
@m-ou-se Unfortunately the dereferencing of raw pointers needs to be done inside the `unsafe` block and it cannot be done in a separate `unsafe` block as we need the variable `value`. I've restored the position of `let ptr = ...` line as it neither requires `catch_unwind` nor `unsafe`. Is this what you had in mind?